### PR TITLE
`ultralytics 8.3.38` SAM 2 video inference

### LINKED ISSUE #14957: SAM2 video predict problem
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I am doing predict on SAM2. This is the demo code
```python
from ultralytics import SAM

# Load a model
model = SAM("sam2_b.pt")

# Display model information (optional)
model.info()

# Segment with bounding box prompt
#results = model("path/to/image.jpg", bboxes=[100, 100, 200, 200])

# Segment with point prompt
results = model("test_video/923_1038.mp4", points=[150, 150])
```
the output is 
```
Model summary: 566 layers, 80,833,666 parameters, 80,833,666 gradients


WARNING ⚠️ inference results will accumulate in RAM unless `stream=True` is passed, causing potential out-of-memory
errors for large sources or long-running streams and videos. See https://docs.ultralytics.com/modes/predict/ for help.

Example:
    results = model(source=..., stream=True)  # generator of Results objects
    for r in results:
        boxes = r.boxes  # Boxes object for bbox outputs
        masks = r.masks  # Masks object for segment masks outputs
        probs = r.probs  # Class probabilities for classification outputs

video 1/1 (frame 1/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 51.8ms
video 1/1 (frame 2/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1741.0ms
video 1/1 (frame 3/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1736.4ms
video 1/1 (frame 4/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1738.8ms
video 1/1 (frame 5/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1740.9ms
video 1/1 (frame 6/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1740.6ms
video 1/1 (frame 7/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1735.1ms
video 1/1 (frame 8/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1738.4ms
video 1/1 (frame 9/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1 10, 1738.9ms
video 1/1 (frame 10/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1 10, 1741.0ms
video 1/1 (frame 11/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1740.2ms
video 1/1 (frame 12/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1737.6ms
video 1/1 (frame 13/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1736.3ms
video 1/1 (frame 14/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1737.9ms
video 1/1 (frame 15/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1738.9ms
video 1/1 (frame 16/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1736.1ms
video 1/1 (frame 17/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1 10, 1736.2ms
video 1/1 (frame 18/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.9ms
video 1/1 (frame 19/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1737.3ms
video 1/1 (frame 20/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1736.2ms
video 1/1 (frame 21/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.9ms
video 1/1 (frame 22/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.4ms
video 1/1 (frame 23/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.3ms
video 1/1 (frame 24/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1736.2ms
video 1/1 (frame 25/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1732.5ms
video 1/1 (frame 26/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1730.9ms
video 1/1 (frame 27/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.4ms
video 1/1 (frame 28/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.5ms
video 1/1 (frame 29/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.8ms
video 1/1 (frame 30/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.5ms
video 1/1 (frame 31/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.2ms
video 1/1 (frame 32/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1731.6ms
video 1/1 (frame 33/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1732.8ms
video 1/1 (frame 34/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.6ms
video 1/1 (frame 35/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1735.5ms
video 1/1 (frame 36/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.1ms
video 1/1 (frame 37/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.9ms
video 1/1 (frame 38/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.1ms
video 1/1 (frame 39/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.8ms
video 1/1 (frame 40/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.0ms
video 1/1 (frame 41/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.7ms
video 1/1 (frame 42/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1735.2ms
video 1/1 (frame 43/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1736.1ms
video 1/1 (frame 44/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.0ms
video 1/1 (frame 45/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.5ms
video 1/1 (frame 46/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.1ms
video 1/1 (frame 47/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.5ms
video 1/1 (frame 48/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.5ms
video 1/1 (frame 49/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.4ms
video 1/1 (frame 50/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1730.3ms
video 1/1 (frame 51/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1729.0ms
video 1/1 (frame 52/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.4ms
video 1/1 (frame 53/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1730.7ms
video 1/1 (frame 54/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1730.4ms
video 1/1 (frame 55/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1728.8ms
video 1/1 (frame 56/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1729.8ms
video 1/1 (frame 57/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1727.0ms
video 1/1 (frame 58/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.0ms
video 1/1 (frame 59/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.5ms
video 1/1 (frame 60/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.1ms
video 1/1 (frame 61/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.7ms
video 1/1 (frame 62/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.6ms
video 1/1 (frame 63/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.8ms
video 1/1 (frame 64/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1730.3ms
video 1/1 (frame 65/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.3ms
video 1/1 (frame 66/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1729.9ms
video 1/1 (frame 67/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1736.1ms
video 1/1 (frame 68/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1735.9ms
video 1/1 (frame 69/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1735.5ms
video 1/1 (frame 70/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.9ms
video 1/1 (frame 71/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.7ms
video 1/1 (frame 72/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.4ms
video 1/1 (frame 73/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1737.3ms
video 1/1 (frame 74/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1736.6ms
video 1/1 (frame 75/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1735.9ms
video 1/1 (frame 76/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.7ms
video 1/1 (frame 77/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.9ms
video 1/1 (frame 78/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1736.7ms
video 1/1 (frame 79/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1737.3ms
video 1/1 (frame 80/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1732.5ms
video 1/1 (frame 81/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1735.7ms
video 1/1 (frame 82/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.1ms
video 1/1 (frame 83/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1734.0ms
video 1/1 (frame 84/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1736.3ms
video 1/1 (frame 85/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1734.3ms
video 1/1 (frame 86/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1731.8ms
video 1/1 (frame 87/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1733.2ms
video 1/1 (frame 88/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.5ms
video 1/1 (frame 89/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.3ms
video 1/1 (frame 90/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.9ms
video 1/1 (frame 91/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.1ms
video 1/1 (frame 92/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.2ms
video 1/1 (frame 93/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.3ms
video 1/1 (frame 94/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.1ms
video 1/1 (frame 95/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1731.0ms
video 1/1 (frame 96/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.7ms
video 1/1 (frame 97/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.3ms
video 1/1 (frame 98/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.1ms
video 1/1 (frame 99/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1733.0ms
video 1/1 (frame 100/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1734.6ms
video 1/1 (frame 101/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1731.3ms
video 1/1 (frame 102/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.5ms
video 1/1 (frame 103/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1734.6ms
video 1/1 (frame 104/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1732.2ms
video 1/1 (frame 105/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1732.8ms
video 1/1 (frame 106/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1730.0ms
video 1/1 (frame 107/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1731.2ms
video 1/1 (frame 108/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1731.5ms
video 1/1 (frame 109/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1739.4ms
video 1/1 (frame 110/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1733.1ms
video 1/1 (frame 111/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1734.0ms
video 1/1 (frame 112/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1737.4ms
video 1/1 (frame 113/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1735.1ms
video 1/1 (frame 114/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1735.8ms
video 1/1 (frame 115/116) /home/dslab/s1101444/SAM2/test_video/923_1038.mp4: 1024x1024 1 0, 1 1, 1 2, 1 3, 1 4, 1 5, 1 6, 1 7, 1 8, 1 9, 1733.5ms
```
I guess only the first frame uses the prompt I selected, and the subsequent frames are segment everything. But shouldn't it be the first frame? After selecting the segment of the object, will this segment continue to be tracked?

question2: What does the parameter labels=[1] represent?

question3: Can predict output the position of the segment?

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Laughing-q**
```python
from ultralytics.models.sam import SAM2VideoPredictor

# Create SAM2VideoPredictor
overrides = dict(conf=0.25, task="segment", mode="predict", imgsz=1024, model="sam2_b.pt", batch=1)
predictor = SAM2VideoPredictor(overrides=overrides)

# Inference with prompts
results = predictor(source="test.mp4", points=[[920, 470], [909, 138]], labels=[1, 1])
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introduced `SAM2VideoPredictor`, a new class for segmenting and tracking objects in videos with enhancements for efficiency and interactivity.

### 📊 Key Changes
- **New Class**: Added `SAM2VideoPredictor` for video interaction and object tracking.
- **Video Segmentation**: Enhanced video segmentation with support for prompts like points and masks.
- **Non-Overlapping Masks**: Integrated functionality to ensure masks don’t overlap.
- **Memory Management**: Improved inference with features like memory clearing and optimization.
- **Documentation Updates**: Expanded documentation for usage examples in video segmentation and tracking.

### 🎯 Purpose & Impact
- **Purpose**:
  - Enhance Ultralytics' segmentation models for dynamic video processing.
  - Provide users with tools to track objects across video frames effectively.
- **Impact**:
  - **For Developers**: Simplifies implementing video-based object segmentation and tracking workflows.
  - **For Users**: Enables robust, real-time object tracking and segmentation with minimal setup.
  - **Usability**: Provides intuitive prompts for video object interactions while maintaining performance. 🚀

### CODE REVIEW

**Comment by Laughing-q** [comment]
@glenn-jocher This PR is ready to review from my side. :)

**Comment by ambitious-octopus** [comment]
@Laughing-q any update here?

**Comment by glenn-jocher** [comment]
@Laughing-q ah I see why the docs compile is not failing, the reference section is only building non-internal methods, so anything starting with an underscore like `_run_single_frame_inference()` is not built as part of the documentation.
Screenshot 2024-11-25 at 16 55 11

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -815,3 +813,795 @@ def get_im_features(self, im):
...
+        Get a dummy object pointer based on an empty mask on the current frame.
+
+        Args:
+            frame_idx (int): The index of the current frame for which to generate the dummy object pointer.
+
+        Returns:
+            torch.Tensor: A tensor representing the dummy object pointer generated from the empty mask.
+        """
+        # Retrieve correct image features
+        current_vision_feats, current_vision_pos_embeds, feat_sizes = self.get_im_features(frame_idx)
```
[the remark this answers is not in the document]
Comment (Laughing-q) [reply ultralytics/models/sam/predict.py]: @glenn-jocher nice catch! fixed in https://github.com/ultralytics/ultralytics/pull/14851/commits/ae79b04cf2b4035417a5e6cbe700e94098408439. Thanks!

**Comment by Laughing-q** [comment]
@glenn-jocher What's the standard of the docstring we are using here? I saw some differences in the `Returns` section between functions:
1. directly use `tuple`:
https://github.com/ultralytics/ultralytics/blob/d3b53d34e5e618e2c002c433e2473abdb800b11f/ultralytics/models/sam/predict.py#L270-L271
2. `tuple` with each element explicitly declared:
https://github.com/ultralytics/ultralytics/blob/d3b53d34e5e618e2c002c433e2473abdb800b11f/ultralytics/models/sam/predict.py#L183-L187
3. directly list all the elements without having a `tuple` on the top:
https://github.com/ultralytics/ultralytics/blob/ee6b36888168382b38827e69a0f902a5f03fc6e9/ultralytics/models/sam/predict.py#L892-L894
4. For `tuple` we are using lowercase but for `List` we are using capital for the first letter:
https://github.com/ultralytics/ultralytics/blob/d3b53d34e5e618e2c002c433e2473abdb800b11f/ultralytics/models/sam/predict.py#L461-L462

**Comment by glenn-jocher** [comment]
@Laughing-q the correct format is here so that MkDocs renders each return correctly.

Good ✅
```
         Returns: 
             (np.ndarray): The output masks in shape CxHxW, where C is the number of generated masks. 
             (np.ndarray): An array of length C containing quality scores predicted by the model for each mask. 
```

 If we use (tuple) and then create a list underneath it then MkDocs only creates 1 table row and stuffs everything inside it:
 
 Bad ❌
 ```
         Returns:
             (tuple) with elements:
                 - (np.ndarray): The output masks in shape CxHxW, where C is the number of generated masks. 
                 - (np.ndarray): An array of length C containing quality scores predicted by the model for each mask. 
```

**Comment by Laughing-q** [comment]
@glenn-jocher Got it! Thank you! I've updated docstring in `sam/predict.py` accordingly.
Also do we want the variable names here? i.e. `new_masks` and `keep`.
https://github.com/ultralytics/ultralytics/blob/b370b0c002eb3b368547fcc3b2b1b3af2b638997/ultralytics/models/sam/predict.py#L587-L589

**Comment by glenn-jocher** [comment]
@Laughing-q yes if we return variables then the names should be there. Sometimes there are no variables though and then we just have the type. Here are the two examples:

```python
def example_function(arg1, arg2=4):
    """
    Example function demonstrating Google-style docstrings.

    Args:
        arg1 (int): The first argument.
        arg2 (int): The second argument, with a default value of 4.

    Returns:
        (bool): True if successful, False otherwise.

    Examples:
        >>> result = example_function(1, 2)  # returns False
    """
    if arg1 == arg2:
        return True
    return False
```

With names
```python
def example_function(arg1, arg2=4):
    """
    Example function demonstrating Google-style docstrings.

    Args:
        arg1 (int): The first argument.
        arg2 (int): The second argument, with a default value of 4.

    Returns:
        result (bool): True if successful, False otherwise.

    Examples:
        >>> result = example_function(1, 2)  # returns False
    """
    result = arg1 == arg2:
    return result
```

**Comment by glenn-jocher** [comment]
@Laughing-q PR merged, nice work on this!

**Pull request merged**